components/feature/session/src/main/java/mozilla/components/feature/session/view/VerticalSwipeRefreshLayout.kt

```diff
@@ -0,0 +1,180 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.session.view
+
+import android.content.Context
+import android.util.AttributeSet
+import android.view.MotionEvent
+import android.view.View
+import android.view.ViewConfiguration
+import androidx.annotation.VisibleForTesting
+import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
+import kotlin.math.abs
+
+/**
+ * [SwipeRefreshLayout] that filters only vertical scrolls for triggering pull to refresh.
+ *
+ * Following situations will not trigger pull to refresh:
+ * - a scroll happening more on the horizontal axis
+ * - a scale in/out gesture
+ * - a quick scale gesture
+ *
+ * To control responding to scrolls and showing the pull to refresh throbber or not
+ * use the [View.isEnabled] property.
+ */
+class VerticalSwipeRefreshLayout @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null
+) : SwipeRefreshLayout(context, attrs) {
+    @VisibleForTesting
+    internal var isQuickScaleInProgress = false
+    @VisibleForTesting
+    internal var quickScaleEvents: Triple<MotionEvent?, MotionEvent?, MotionEvent?> = Triple(null, null, null)
```

Comment:
This should be either a data class or just 3 local variables. The lack of naming makes the code hard to read. If you go the data class route, you may want to use mutable `var`s.

Reply:
Fair. Will use a data class. Thanks!